agrest-engine/src/main/java/io/agrest/exp/parser/ExpScalar.java

```diff
@@ -31,22 +39,44 @@ protected Object transformExpression(Function<Object, Object> transformer) {
      * Accept the visitor.
      **/
     public <T> T jjtAccept(AgExpressionParserVisitor<T> visitor, T data) {
-
         return visitor.visit(this, data);
     }
 
+    public String getScalarImage() {
+        return scalarImage;
+    }
+
+    public void setScalarImage(String scalarImage) {
+        this.scalarImage = scalarImage;
+    }
+
     @Override
     protected ExpScalar shallowCopy() {
         return new ExpScalar(getValue());
     }
 
     @Override
     public String toString() {
+        if (value == null) {
+            return "null";
+        }
         if (value instanceof CharSequence) {
-            return "'" + value + "'";
-        } else {
-            return String.valueOf(value);
+            String quoteChar = scalarImage.substring(0, 1);
```

Comment:
Let's try to move this logic to the setter, so here it will simply return `scalarImage`